sklearn/hpsearch/gke_parallel.py

```diff
@@ -0,0 +1,437 @@
+
+        print('Deploying worker {}'.format(worker_id))
+        create_job(job_body)
+
+
+    def _partition_grid(self, param_grid_dict, partition_keys):
+        _param_grid_dict = deepcopy(param_grid_dict)
+
+        partition_lists = [_param_grid_dict.pop(key) for key in partition_keys]
+        
+        partitioned = []
+        for prod in product(*partition_lists):
+            lists = [[element] for element in prod]
+            singleton = dict(zip(partition_keys, lists))
+            singleton.update(_param_grid_dict)
+
+            partitioned.append(singleton)
+
+        return partitioned
+
+
+    def _partition_param_grid(self, param_grid, target_n_partition=5):
+        """Returns a list of param_grids whose union is the input
+        param_grid.
+
+        If param_grid is a dict:
+
+        The implemented strategy attempts to partition the param_grid
+        into at least target_n_partition smaller param_grids.
+
+        NOTE: The naive strategy implemented here does not distinguish
+        between different types of parameters nor their impact on the
+        running time.  The user of this module is encouraged to
+        implement their own paritioning strategy based on their needs.
+        """
+        if type(param_grid) == list:
+            # If the input is already a list of param_grids then just
+            # use it as is.
+            return param_grid
+        else:
+            # The strategy is to simply expand the grid fully with
+            # respect to a parameter:
+            # [1, 2, 3]x[4, 5] --> [1]x[4, 5], [2]x[4, 5], [3]x[4, 5]
+            # until the target number of partitions is reached.
+            partition_keys = []
+            n_partition = 1
+            for key, lst in param_grid.items():
+                partition_keys.append(key)
+                n_partition *= len(lst)
+
+                if n_partition >= target_n_partition:
+                    break
+
+            partitioned = self._partition_grid(param_grid, partition_keys)
+
+            return partitioned
+
+
+    def _handle_grid_search(self, X_uri, y_uri):
+        param_grids = self._partition_param_grid(self.search.param_grid, self.n_nodes)
+
+        for i, param_grid in enumerate(param_grids):
+            worker_id = str(i)
+
+            self.param_grids[worker_id] = param_grid
+            self.job_names[worker_id] = self._make_job_name(worker_id)
+            self.output_uris[worker_id] = 'gs://{}/{}/{}/fitted_search.pkl'.format(self.bucket_name, self.task_name, worker_id)
+            self.output_without_estimator_uris[worker_id] = 'gs://{}/{}/{}/fitted_search_without_estimator.pkl'.format(self.bucket_name, self.task_name, worker_id)
+            self.dones[worker_id] = False
+
+            pickle_and_upload(param_grid, self.bucket_name, '{}/{}/param_grid.pkl'.format(self.task_name, worker_id))
+
+            # TODO: Make sure that each job is deployed to a different node.
```

Comment:
TODO implemented?

Reply:
Not yet - need to implement round-robin across the nodes.

Reply:
thinking about this again, I might move this (along with other TODOs) into an issue and we can evaluate later.  I suspect most training tasks will not really benefit from multiple nodes.  e.g. a single node of 64 cores is better than 2 nodes with 32 cores each.